File: models/Gpt-3.5-turbo/cpp/code/problems/string_algorithms/Longest_Palindrome.cpp

#include <iostream>
#include <vector>
using namespace std;

string longestPalindrome(string s) {
    int n = s.length();
    vector<vector<bool>> dp(n, vector<bool>(n, false));

    // Single characters are palindromes
    for (int i = 0; i < n; i++) {
        dp[i][i] = true;
    }

    int maxLen = 1, start = 0;

    // Check for palindromes of length 2 or more
    for (int length = 2; length <= n; length++) {
        for (int i = 0; i < n - length + 1; i++) {
            int end = i + length - 1;

            if (s[i] == s[end]) {
                // If inner substring is palindrome, then current substring is also palindrome
                if (length == 2 || dp[i + 1][end - 1]) {
                    dp[i][end] = true;

                    if (length > maxLen) {
                        maxLen = length;
                        start = i;
                    }
                }
            }
        }
    }

    return s.substr(start, maxLen);
}

int main() {
    string s;
    cin >> s;

    cout << longestPalindrome(s) << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/string_algorithms/Longest_Palindrome.py

def longest_palindrome(s):
    n = len(s)
    start = end = 0

    for i in range(n):
        # Check for odd-length palindromes
        left = right = i
        while left >= 0 and right < n and s[left] == s[right]:
            if right - left + 1 > end - start + 1:
                start = left
                end = right
            left -= 1
            right += 1

        # Check for even-length palindromes
        left = i
        right = i + 1
        while left >= 0 and right < n and s[left] == s[right]:
            if right - left + 1 > end - start + 1:
                start = left
                end = right
            left -= 1
            right += 1

    return s[start:end+1]
